### DISCUSSION TITLE
--reflog option for git log

### DISCUSSION THREAD

**Discussion by guillaumeboehm** [magit/magit#4748] [Ideas]
Any thoughts on adding the `--reflog` option for `git log` to spot orphan commits in a nice little graph ? With a shortcut like `-R` maybe, or directly `l R` ?

**Comment by kyleam** [comment]
I'm not sure about whether Magit itself should support it, but fwiw you can add such a command yourself relatively easily:

```elisp
(defun my/magit-log-reflog (&optional args files)
  "Show log for objects mentioned in reflog."
  (interactive (magit-log-arguments))
  (magit-log-setup-buffer (list "--reflog") args files))

(transient-append-suffix 'magit-log "o"
  '("R" "reflog objects" my/magit-log-reflog))
```

That adds the command to the "Log" column, but you could of course move it under "Reflog" if you prefer.

**Reply by guillaumeboehm** [reply]
> I'm not sure about whether Magit itself should support it

Yeah I agree it's quite specific, but since reflog commands are already available I thought maybe that's reasonable.

> but fwiw you can add such a command yourself relatively easily

Oh that's really nice, thanks! I will definitely use this :)

**Reply by tarsius** [reply]
I wasn't aware this existed but now that I know it exists I could see myself using it occasionally if it were readily available.

What does `--reflog` without additional arguments show? Just the `HEAD` reflog or those of all branches? In my case the latter would probably be very noisy and thus useless. But showing the reflog for the current branch only, might be useful.

Note sure whether it would make more sense to add the argument as an infix command or to add something like `magit-log-current-with-reflog`. I would probably only ever want to use this for the current branch, so it would makes sense to add a command for that as it is less key presses, but others might want to use it for a branch that isn't checked out, and adding multiple such commands would be a bit much, so I am leaning towards adding just the argument. (But where to place it?)

In any case the addition should be hidden by default.

**Reply by guillaumeboehm** [reply]
> What does --reflog without additional arguments show? Just the HEAD reflog or those of all branches? In my case the latter would probably be very noisy and thus useless. But showing the reflog for the current branch only, might be useful.

Hum, I only use --all so I assumed that only the HEAD would be used otherwise but it seems that all the branches are shown just with the --reflog option (I might be wrong I'm still a bit new to advanced git features).

**Reply by kyleam** [reply]
Right, it prepends all objects mentioned in the reflog.  It's not just HEAD's.

**Reply by guillaumeboehm** [reply]
> Right, it prepends all objects mentioned in the reflog. It's not just HEAD's.

Yeah ok, well I still feel like the option could be in magit by default.
If you use `log --graph --oneline --reflog` it allows you to find orphan commits much more easily than by scanning the reflog by hand (I believe they necessarily stick out from the graph so maybe even with a bloated tree?).
Although I completely agree that it does not have to be visible, the use case is so specific it is maybe even better as just a command.

**Reply by guillaumeboehm** [reply]
Hi, sorry I'm reviving the thread after a while. I'm trying to use the command you provided @kyleam but I get that `magit-log is not a transient command`. From what I could find this seems to be a magit version issue most of the time, but I'm a bit confused about my version of magit. Running `M-x magit-version` returns the commit hash `c1fb53d` instead of a version number and I couldn't find the hash in the commit history. I'm using doom emacs which handles magit, so I'm not specifying it myself...

**Reply by kyleam** [reply]
@guillaumeboehm I don't know anything about doom, but perhaps the issue is just that you haven't loaded magit yet (that matches the error you would see).  Try delaying the snippets above until magit is available with something like ` (with-eval-after-load 'magit ...)`.

**Reply by guillaumeboehm** [reply]
Okay! I see, that works perfectly thank you very much @kyleam !

**Reply by tarsius** [reply]
I've added `magit-log-reflog` but put it on level 0, which means that it has to be enabled individually. (307abb6002257045d1cd168a92c799d84bc98dbf)

I think `--reflog` has potential, but without the ability to limit to the reflog of a specific branch, it cannot live up to that.
